examples/adjustwindow_updateviewer_saveimage.py

```diff
@@ -0,0 +1,68 @@
+from napari.settings import get_settings
+import time
+import napari
+from napari._qt.qthreading import thread_worker
+from skimage import data
+from skimage.morphology import ball, octahedron
+import matplotlib.pyplot as plt
+
+
+def make_screenshot(viewer):
+    img = viewer.screenshot(canvas_only=True, flash=False)
+    plt.imshow(img)
+    plt.axis("off")
+    plt.show()
+
+
+get_settings().application.save_window_state = False
+get_settings().application.window_size = (500, 500)  # h, w
+get_settings().application.window_position = (900, 300)  # h, w
+viewer = napari.Viewer()
+viewer.window.qt_viewer.dockLayerControls.toggleViewAction().trigger()
+# in napari 0.4.13 it will become private:
+# viewer.window._qt_viewer.dockLayerControls.toggleViewAction().trigger()
+
```

Comment:
0.4.13 is about to come out (we're currently testing release candidates), so we should definitely avoid suggesting the use of a private attribute in an example! Why is this private call necessary?

Reply:
Sorry if this came out a bit dry, I meant it as part as a wider review and it fired off as a single comment :) I'll add more when I understood the whole thing!

Reply:
> 0.4.13 is about to come out (we're currently testing release candidates), so we should definitely avoid suggesting the use of a private attribute in an example! Why is this private call necessary?

See related discussion: https://forum.image.sc/t/hide-layer-controls-panel-from-python/61975/8?u=haesleinhuepf

Reply:
Hah, it's been discussed already! Thanks for the link @haesleinhuepf. Ok then! :+1:

Reply:
> Why is this private call necessary?

Essentially, we're trying to make *all* of the actual Qt API a napari application implementation detail, and not part of our public API.  Of course, there's a ton of things an end-user might want to do _with_ Qt (and this example is using a Qt API directly).

It's a difficult thing to decide and will definitely require some discussion in the future.  If we just say "go ahead, use any of our Qt objects, here's the Qt API", then people can do anything, and that's great, but we can't evolve at all without breaking a ton of stuff.

If we make it private, people at least know that we're not guaranteeing we won't change it (and that plugins definitely shouldn't use it).  But end-users can still access "riskily" :)

I don't have a super strong feeling for a "rule" on how we should handle this, so, for now, I tend to just say "here's how ya do it, but, just a heads up here...".  And if we get a couple requests to make a public API for something, we'll know it's important.  (on the other hand, we don't want to just forward the entire Qt API, so some decision making is necessary)

Reply:
From my user perspective:
> but we can't evolve at all without breaking a ton of stuff.

I think breaking stuff to evolve is a good strategy, adjusting my user scripts to new version is not a big deal.

Regarding this special case of showing and hiding the side panel from python side:
I am mostly using napari on my laptop with split-screen: left VSCode, right napari.
As there is only limited space, I want to see more of my data and switch off the side panel manually.
Next,  I change some settings from python, e.g.
```py
viewer.camera.zoom = 1.3
viewer.axes.visible = True
viewer.camera.angles = (90, 27, 145)
```
But these parameters are optimized for the case that the side panel is off.
As soon as I close my program and start it again a few days later, napari will start with the side panel on, hence the viewer has a different aspect ratio in the canvas and therefore my settings that I've set last time won't work anymore.
Of course, I can simply remember that I made these settings with "side panel is off", but then I would have to remember for all my scripts, if I optimized the viewer parameter with  "side panel is off" or  "side panel on".
That is why I think it would be nice if there was an option for users to change this setting by the API.

Reply:
> I think breaking stuff to evolve is a good strategy,

I am also someone who doesn't mind updating my scripts when a dependency makes breaking changes, but I think we might be in the minority there.  Or, at the very least, it's not considered a good practice for a package to make unannounced breaking changes.  With public attributes and methods, we pretty much _must_ go through a deprecation process: announcing the change and then later making it.  It makes for a _slower_ change process, and a messier code base.  Private methods announce: "here is something that we reserve the right to change without notice".  If you are the kind of person who doesn't mind updating stuff when it changes and are happy for your libs to make breaking changes, then private methods will work nicely for you :)

Reply:
one more thought: this particular case brings up what might be a good rule of thumb.  Anything that has such an easy thing to do via the menus (here: toggling an item in the `Window` menu) should probably have a corresponding public method

Reply:
One of the first things in the GitHub README is:
> We're developing napari in the open! But the project is in an **alpha** stage, and there will still likely be **breaking changes** with each release. 

So even when there is no announcement that this method will become private, this breaking change is in my opinion completely fine. 
I like your philosophy of private methods, that is a nice middle course: Giving the developers more freedom for faster development while still preserving the freedom of users to tweak also internal properties.

Regarding your second idea: public methods for Window property would be amazing!
This line was also a bit hacky:
```py
from napari.settings import get_settings
get_settings().application.window_size = (500, 500)
```
Maybe this can also become a public method?

Additional thought: Changing `window_size` has to happen *before* the viewer is initialized, while hiding the side panel has to happen *after* the viewer is initialized. Maybe this can be more coherent, maybe everything can happen after the viewer is initialized?

Reply:
> from napari.settings import get_settings; get_settings().application.window_size = (500, 500)
> Maybe this can also become a public method?

that _is_ a public method :)  anything without a leading underscore is public.

Reply:
>  Changing window_size has to happen before the viewer is initialized, while hiding the side panel has to happen after the viewer is initialized.

only if you change it in the settings, afterwards, you'd need to use `viewer.window.resize(width, height)`

Reply:
> `viewer.window.resize(width, height)`

This is exactly what I was looking for 😍
I'll change this pr accordingly in a few minutes.
I just realized that there is a minimum size that is required.
Maybe one can send the user a warning when the user wants to set one value below a certain threshold here?
https://github.com/napari/napari/blob/1cc0592a02bd7b06c3b01c16128466aea9868494/napari/_qt/qt_main_window.py#L994  I've tried a few values and the minimum value seems to be somewhere at viewer.window.resize(600, 550)